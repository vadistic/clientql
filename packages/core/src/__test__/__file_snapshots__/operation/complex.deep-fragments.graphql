query FindEventsAtVenueMyQuery($venueId: ID!) {
  findEventsAtVenue(venueId: $venueId) {
    ...EventDeep
  }
}

fragment EventDeep on Event {
  id
  name
  startsAt
  endsAt
  venue {
    ...VenueDeep
  }
  minAgeRestriction
  ... on Concert {
    ...ConcertDeep
  }
  ... on Festival {
    ...FestivalDeep
  }
  ... on Conference {
    ...ConferenceDeep
  }
}

fragment VenueDeep on Venue {
  __typename
  id
  name
  address
  maxOccupancy
}

fragment ConcertDeep on Concert {
  __typename
  id
  name
  startsAt
  endsAt
  venue {
    ...VenueDeep
  }
  previousVenues {
    ...VenueDeep
  }
  minAgeRestriction
  performingBand {
    ...PerformerDeep
  }
}

fragment PerformerDeep on Performer {
  __typename
  id
  name
  contact
  speciality
  fee
}

fragment FestivalDeep on Festival {
  __typename
  id
  name
  startsAt
  endsAt
  venue {
    ...VenueDeep
  }
  minAgeRestriction
  performers {
    ...PerformerDeep
  }
}

fragment ConferenceDeep on Conference {
  __typename
  id
  name
  startsAt
  endsAt
  venue {
    ...VenueDeep
  }
  minAgeRestriction
  speakers {
    ...SpeakerDeep
  }
  workshops
}

fragment SpeakerDeep on Speaker {
  __typename
  id
  name
  contact
}
