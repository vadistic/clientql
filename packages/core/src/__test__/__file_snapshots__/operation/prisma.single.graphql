query UserQuery($where: UserWhereUniqueInput!) {
  user(where: $where) {
    ...UserFlatPartial
    posts {
      ...PostFlatPartial
    }
  }
}

fragment UserFlatPartial on User {
  __typename
  id
  nationality
  email
  name
}

fragment PostFlatPartial on Post {
  __typename
  id
  content
  createdAt
  published
  title
  updatedAt
}
