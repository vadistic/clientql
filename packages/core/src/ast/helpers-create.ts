import {
  ArgumentNode,
  FieldNode,
  FragmentDefinitionNode,
  FragmentSpreadNode,
  InlineFragmentNode,
  InputValueDefinitionNode,
  NameNode,
  OperationDefinitionNode,
  OperationTypeNode,
  SelectionNode,
  VariableDefinitionNode,
} from 'graphql'
import { Kind } from './kind'
import { Typename } from './types'

export const createName = (name: string): NameNode => ({
  kind: Kind.NAME,
  value: name,
})

export interface CreateFieldProps {
  fieldname: string
  arguments?: ArgumentNode[]
  selections?: SelectionNode[]
}

export const createField = (props: CreateFieldProps): FieldNode => ({
  kind: Kind.FIELD,
  name: createName(props.fieldname),
  arguments: props.arguments,
  selectionSet: props.selections
    ? {
        kind: Kind.SELECTION_SET,
        selections: props.selections,
      }
    : undefined,
})

export interface CreateFragmentProps {
  fragmentname: string
  condition: Typename
  selections: SelectionNode[]
}
/**
 *  no fragment variables (yet?)
 */
export const createFragment = ({
  condition,
  selections,
  fragmentname,
}: CreateFragmentProps): FragmentDefinitionNode => ({
  kind: Kind.FRAGMENT_DEFINITION,
  name: createName(fragmentname),
  typeCondition: {
    kind: Kind.NAMED_TYPE,
    name: createName(condition),
  },
  selectionSet: {
    kind: Kind.SELECTION_SET,
    selections,
  },
})

export const createFragmentSpread = (name: string): FragmentSpreadNode => ({
  kind: Kind.FRAGMENT_SPREAD,
  name: createName(name),
})

export const createInlineFragment = (
  on: Typename,
  selections: SelectionNode[],
): InlineFragmentNode => ({
  kind: Kind.INLINE_FRAGMENT,
  typeCondition: {
    kind: Kind.NAMED_TYPE,
    name: createName(on),
  },
  selectionSet: {
    kind: Kind.SELECTION_SET,
    selections,
  },
})

export const createOperationVariable = (suffix: string) => (
  node: InputValueDefinitionNode,
): VariableDefinitionNode => ({
  kind: Kind.VARIABLE_DEFINITION,
  defaultValue: node.defaultValue,
  directives: node.directives,
  variable: {
    kind: Kind.VARIABLE,
    name: createName(node.name.value + suffix),
  },
  type: node.type,
})

export const createOperationArgument = (suffix: string) => (
  node: InputValueDefinitionNode,
): ArgumentNode => ({
  kind: Kind.ARGUMENT,
  name: node.name,
  value: {
    kind: Kind.VARIABLE,
    name: createName(node.name.value + suffix),
  },
})

export interface CreateOperationProps {
  name: string
  type: OperationTypeNode
  variables?: VariableDefinitionNode[]
  selections: SelectionNode[]
}

export const createOperation = (props: CreateOperationProps): OperationDefinitionNode => ({
  kind: Kind.OPERATION_DEFINITION,
  name: createName(props.name),
  operation: props.type,
  selectionSet: {
    kind: Kind.SELECTION_SET,
    selections: props.selections,
  },
  variableDefinitions: props.variables,
})
