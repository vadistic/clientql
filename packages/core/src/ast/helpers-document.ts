import { ASTKindToNode, DefinitionNode, DocumentNode, KindEnum, SelectionNode } from 'graphql'
import { truthy } from '../utils/types'
import { Kind } from './kind'

/**
 * created doc node fromd efinition/s
 */

export const wrapDocument = (
  ...nodes: Array<DefinitionNode | undefined | null | false>
): DocumentNode => ({
  kind: Kind.DOCUMENT,
  definitions: nodes.filter(truthy),
})

/**
 * get mutable arr of definitions from doc node
 */

export const unwrapDocument = (doc: DocumentNode): DefinitionNode[] => [...doc.definitions]

/**
 * get first/firstOfKind definition from docuemnt
 */
export interface GetDocDefinition {
  <K extends KindEnum>(doc: DocumentNode, kind: K): ASTKindToNode[K] | undefined
  (doc: DocumentNode): DefinitionNode
}

export const getDocDefinition: GetDocDefinition = <K extends KindEnum>(
  doc: DocumentNode,
  kind?: K,
) => {
  if (!doc.kind || doc.kind !== Kind.DOCUMENT || !Array.isArray(doc.definitions)) {
    throw Error('getDocDefinition: Invalid Document (like no document at all')
  }

  if (kind) {
    for (const node of doc.definitions) {
      if (node.kind === kind) {
        return node
      }
    }

    // this should be automatic...
    return undefined as any
  }

  return doc.definitions[0]
}

// for interfaces/unions that need it flat
export const unwrapSelectionSet = (node?: SelectionNode) =>
  (node && node.kind === Kind.FIELD && node.selectionSet && [...node.selectionSet.selections]) ||
  undefined
