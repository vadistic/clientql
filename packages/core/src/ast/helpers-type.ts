import { ListTypeNode, NamedTypeNode, TypeNode } from 'graphql'
import { Kind, KindMap } from './kind'

/**
 * alternative to those nested Null/List types that is easier to print
 */
export type TypeModifier = KindMap['NON_NULL_TYPE'] | KindMap['LIST_TYPE']

export interface UnwrapType {
  type: NamedTypeNode
  typename: string
  modifiers: TypeModifier[]
}

export const unwrapType = (
  node: TypeNode,
  modifiers: TypeModifier[] = [],
): UnwrapType => {
  if (node.kind === Kind.NON_NULL_TYPE) {
    return unwrapType(node.type, [...modifiers, Kind.NON_NULL_TYPE])
  }

  if (node.kind === Kind.LIST_TYPE) {
    return unwrapType(node.type, [...modifiers, Kind.LIST_TYPE])
  }

  // node.kind === Kind.NAMED_TYPE
  return {
    type: node,
    typename: node.name.value,
    modifiers,
  }
}

/**
 * same as unwarpType but without those modifiers when not needed
 */

export const getTypename = (node: TypeNode): string =>
  node.kind === Kind.NAMED_TYPE ? node.name.value : getTypename(node.type)

/**
 * means top level value is nullable
 */
export const isNullable = (
  node: TypeNode,
): node is ListTypeNode | NamedTypeNode => node.kind !== Kind.NON_NULL_TYPE

/**
 * recursively check if field type will be a list
 */

export const isList = (node: TypeNode): boolean =>
  node.kind === Kind.LIST_TYPE
    ? true
    : node.kind === Kind.NAMED_TYPE
    ? false
    : isList(node.type)
