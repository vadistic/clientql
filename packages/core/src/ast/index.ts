export * from './helpers-type'
export * from './helpers-document'
export * from './helpers-create'

export * from './kind'
export * from './merge-extensions'
export * from './guards'
export * from './types'
