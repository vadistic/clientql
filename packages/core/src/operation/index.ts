export * from './selection'
export * from './operation'
export * from './utils'
