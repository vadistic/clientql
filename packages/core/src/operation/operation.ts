import {
  ArgumentNode,
  FieldNode,
  FragmentDefinitionNode,
  SelectionNode,
  VariableDefinitionNode,
} from 'graphql'
import {
  createField,
  createOperation,
  createOperationArgument,
  createOperationVariable,
  Fieldname,
  Typename,
  unwrapSelectionSet,
  wrapDocument,
} from '../ast'
import { CoreProps } from '../config'
import { capitalise } from '../utils'
import { buildSelection } from './selection'
import { findRootOperation, retriveCacheFragments } from './utils'

/**
 * this needs to build operation from arr of fieldnames
 */

export const buildOperation = (props: CoreProps) => (path: Fieldname[]) => {
  const [head] = path
  const root = findRootOperation(props)(head)

  // head not on any of roots
  if (!root) {
    return
  }

  const { fragments, selections, variables } = buildDeepOperation(props)(
    root.name,
    path,
  )

  const operationName = path.map(capitalise).join('') + capitalise(root.type)

  const operation = createOperation({
    name: operationName,
    selections,
    type: root.type,
    variables,
  })

  return wrapDocument(operation, ...fragments)
}

interface BuildDeepOperation {
  variables: VariableDefinitionNode[]
  fragments: FragmentDefinitionNode[]
  selections: SelectionNode[]
}

const buildDeepOperation = (props: CoreProps) => (
  target: Typename,
  path: Fieldname[],
  level = 0,
): BuildDeepOperation => {
  const [head, ...tail] = path

  const vtx = props.graph.get(target)

  if (!vtx) {
    throw Error('Invalid operation target')
  }

  const edgeData = vtx.weightsMap!.get(head)
  const nextTarget = vtx.edgesMap!.get(head)

  if (!edgeData || !nextTarget) {
    throw Error('Invalid operation path')
  }

  const argSuffix = level > 0 ? '' + level : ''

  const { variables, args } = (edgeData.arguments || []).reduce(
    (acc, input) => {
      acc.args.push(createOperationArgument(argSuffix)(input))
      acc.variables.push(createOperationVariable(argSuffix)(input))
      return acc
    },
    {
      variables: [] as VariableDefinitionNode[],
      args: [] as ArgumentNode[],
    },
  )

  // base
  if (tail.length === 0) {
    const { selection, fragmentnames } = buildSelection(props)([
      [head, nextTarget],
    ])

    const field = createField({
      arguments: args,
      fieldname: head,
      selections: unwrapSelectionSet(selection),
    })

    return {
      fragments: retriveCacheFragments(props)(fragmentnames),
      variables,
      selections: [field],
    }
  } else {
    // well, not a base
    const {
      selections,
      fragments,
      variables: deepVariables,
    } = buildDeepOperation(props)(nextTarget, tail, level + 1)

    const nextSelection: FieldNode = createField({
      fieldname: head,
      arguments: args,
      selections,
    })

    return {
      variables: [...variables, ...deepVariables],
      fragments,
      selections: [nextSelection],
    }
  }
}
