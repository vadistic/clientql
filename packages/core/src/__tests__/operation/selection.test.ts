import { print } from 'graphql'
import { createField, createFragment, wrapDocument } from '../../ast'
import { buildSelection } from '../../operation'
import { complexProps, prismaProps } from '../fixture'

describe('selection', () => {
  it('not very nested prisma fragment', () => {
    const selectionFn = buildSelection(prismaProps)

    const { selection, fragmentnames } = selectionFn([[null as any, 'Post']])

    const frag = createFragment({
      fragmentname: 'PostFragment',
      selections: [...selection!.selectionSet!.selections!],
      condition: 'Post',
    })

    expect(print(wrapDocument(frag))).toMatchSnapshot()
  })

  it('complex fragment selection no interface', () => {
    const selectionFn = buildSelection(complexProps)

    const { selection, fragmentnames } = selectionFn([[null as any, 'Event']])

    const frag = createFragment({
      fragmentname: 'EventFragment',
      condition: 'Event',
      selections: [
        createField({
          fieldname: 'event',
          selections: [...selection!.selectionSet!.selections!],
        }),
      ],
    })

    console.log(fragmentnames)

    expect(print(wrapDocument(frag))).toMatchSnapshot()
  })
})
