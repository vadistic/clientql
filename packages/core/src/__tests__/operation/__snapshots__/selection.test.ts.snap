// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selection complex fragment selection no interface 1`] = `
"fragment EventFragment on Event {
  event {
    ...EventFlat
    venue {
      ...VenueFlat
    }
    ... on Concert {
      ...ConcertFlat
      venue {
        ...VenueFlat
      }
      previousVenues {
        ...VenueFlat
      }
      performingBand {
        ...PerformerFlat
      }
    }
    ... on Festival {
      ...FestivalFlat
      venue {
        ...VenueFlat
      }
      performers {
        ...PerformerFlat
      }
    }
    ... on Conference {
      ...ConferenceFlat
      venue {
        ...VenueFlat
      }
      speakers {
        ...SpeakerFlat
      }
    }
  }
}
"
`;

exports[`selection not very nested prisma fragment 1`] = `
"fragment PostFragment on Post {
  ...PostFlat
  author {
    ...UserFlat
  }
}
"
`;
